plugins/modules/dcnm_image_upgrade.py

```diff
@@ -0,0 +1,5313 @@
+            version, e.g. "12.1.2e"
+            None
+        """
+        return self._get("version")
+
+    @property
+    def version_major(self):
+        """
+        Return the NDFC major version, if it exists.
+        Return None otherwise
+
+        We are assuming semantic versioning based on:
+        https://semver.org
+
+        Possible values:
+            if version is 12.1.2e, return 12
+            None
+        """
+        if self.version is None:
+            return None
+        return (self._get("version").split("."))[0]
+
+    @property
+    def version_minor(self):
+        """
+        Return the NDFC minor version, if it exists.
+        Return None otherwise
+
+        We are assuming semantic versioning based on:
+        https://semver.org
+
+        Possible values:
+            if version is 12.1.2e, return 1
+            None
+        """
+        if self.version is None:
+            return None
+        return (self._get("version").split("."))[1]
+
+    @property
+    def version_patch(self):
+        """
+        Return the NDFC minor version, if it exists.
+        Return None otherwise
+
+        We are assuming semantic versioning based on:
+        https://semver.org
+
+        Possible values:
+            if version is 12.1.2e, return 2e
+            None
+        """
+        if self.version is None:
+            return None
+        return (self._get("version").split("."))[2]
+
+
+def main():
+    """main entry point for module execution"""
+
+    element_spec = dict(
+        config=dict(required=True, type="dict"),
+        state=dict(default="merged", choices=["merged", "deleted", "query"]),
+    )
+
+    module = AnsibleModule(argument_spec=element_spec, supports_check_mode=True)
+    dcnm_module = NdfcAnsibleImageUpgrade(module)
+    dcnm_module.validate_input()
+    dcnm_module.get_have()
+    dcnm_module.get_want()
+
+    if module.params["state"] == "merged":
+        dcnm_module.get_need_merged()
+    elif module.params["state"] == "deleted":
+        dcnm_module.get_need_deleted()
+    elif module.params["state"] == "query":
+        dcnm_module.get_need_query()
+
+    if module.params["state"] == "query":
+        dcnm_module.result["changed"] = False
+    if module.params["state"] in ["merged", "deleted"]:
+        if dcnm_module.need:
+            dcnm_module.result["changed"] = True
+        else:
+            module.exit_json(**dcnm_module.result)
+
+    if module.check_mode:
+        dcnm_module.result["changed"] = False
+        module.exit_json(**dcnm_module.result)
+
+    if dcnm_module.need:
+        if module.params["state"] == "merged":
+            dcnm_module.handle_merged_state()
+        elif module.params["state"] == "deleted":
+            dcnm_module.handle_deleted_state()
+        elif module.params["state"] == "query":
+            dcnm_module.handle_query_state()
+
+    module.exit_json(**dcnm_module.result)
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
General comment. I have gone through the module code and have added comments. But would like to understand the general flow for each state. Would be good if you can spare sometime next week so that we can have a sit in review. Include Praveen Ramamoorthy and Mike also for this review. Check with Mike and schedule a call. We are from India and so please select appropriate time that suits all.